### PR TITLE
[WIP] fix #1700

### LINKED ISSUE #1700: Fault tolerant supports cluster-host level, not only host level
### The New Feature

Fault tolerant supports cluster-host level, not only host level.

### Your scenes

Cluster A and cluster B share host 1, but the two clusters need independent fault tolerance.

### Your advice

Describe the advice or solution of this new feature.

### Environment

- 0.20.0

### PR DESCRIPTION
**Pull request by taoyuanyuan**
### Issues associated with this PR

Your PR should present related issues you want to solve.

### Solutions
#1700 

### UT result
Unit Test is needed if the code is changed, your unit test should cover boundary cases, corner cases, and some exceptional cases. And you need to show the UT result.

### Benchmark
If your code involves the processing of every request, you should give the Benchmark Result.

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

### CODE REVIEW

File: pkg/proxy/downstream.go [hunk 1]
```diff
@@ -841,15 +841,15 @@ func (s *downStream) chooseHost(endStream bool) {
...
+	pool, host, err := s.initializeUpstreamConnectionPool(s)
 	if err != nil {
 		log.Proxy.Alertf(s.context, types.ErrorKeyUpstreamConn, "initialize Upstream Connection Pool error, request can't be proxyed, error = %v", err)
 		s.requestInfo.SetResponseFlag(api.NoHealthyUpstream)
 		s.sendHijackReply(api.NoHealthUpstreamCode, s.downstreamReqHeaders)
 		return
 	}
-	s.requestInfo.OnUpstreamHostSelected(pool.Host())
-	s.requestInfo.SetUpstreamLocalAddress(pool.Host().AddressString())
+	s.requestInfo.OnUpstreamHostSelected(host)
```
Comment (nejisama) [inline pkg/proxy/downstream.go:851]: 这里不设置在这里吧，之前讨论的是在cluster manager下设置，把这里的设置去掉

Comment (taoyuanyuan) [reply pkg/proxy/downstream.go]: 收口到 initializeUpstreamConnectionPool 里了

File: pkg/upstream/cluster/host.go [hunk 2]
```diff
@@ -56,7 +56,8 @@ func NewSimpleHost(config v2.Host, clusterInfo types.ClusterInfo) types.Host {
 		metaData:      config.MetaData,
 		tlsDisable:    config.TLSDisable,
 		weight:        config.Weight,
-		healthFlags:   GetHealthFlagPointer(config.Address),
+		//healthFlags:   GetHealthFlagPointer(config.Address),
+		healthFlags:   new(uint64),
```
Comment (nejisama) [inline pkg/upstream/cluster/host.go:60]: 不仅仅是要改这里，Host所有Flag相关的地方都要改

File: pkg/upstream/cluster/cluster_manager.go [hunk 3]
```diff
@@ -447,13 +449,13 @@ func (cm *clusterManager) getActiveConnectionPool(balancerContext types.LoadBala
 				continue
 			}
 			if pools[i].CheckAndInit(balancerContext.DownstreamContext()) {
-				return pools[i], nil
+				return pools[i], hosts[i], nil
 			}
 		}
 		waitTime := tryConnTimes[t]
 		time.Sleep(waitTime)
 	}
```
Comment (nejisama) [inline pkg/upstream/cluster/cluster_manager.go:457]: 应该不用改接口，直接在这里设置Host就可以了

Comment (nejisama) [reply pkg/upstream/cluster/cluster_manager.go:457]: 想了一下，其实这个地方改不改 对于故障隔离这个需求影响不大。 故障隔离的Cluster没有基于Host获取，而是根据Router获取的

Comment (taoyuanyuan) [reply pkg/upstream/cluster/cluster_manager.go:457]: 但故障隔离需要通过 requestinfo获取 Host

Comment (taoyuanyuan) [reply pkg/upstream/cluster/cluster_manager.go:457]: 感觉还是在proxy里面设置更合理，cluster不需要感知这个requetinfo

File: pkg/upstream/cluster/host.go (hunk 2 above)
Comment (taoyuanyuan) [reply pkg/upstream/cluster/host.go:60]: 还有啥？

Comment (nejisama) [reply pkg/upstream/cluster/host.go:60]: Set/Clear/Get 都要改哦

Comment (taoyuanyuan) [reply pkg/upstream/cluster/host.go:60]: 不用改吧 每个host都是独立的，对方法是透明的

**Pull request closed without merging**